judge/views/user.py

```diff
@@ -227,6 +233,94 @@ def get(self, request, *args, **kwargs):
+    @cached_property
+    def data_path(self):
+        return os.path.join(settings.DMOJ_USER_DATA_CACHE, '%s.zip' % self.request.profile.id)
+
+    def dispatch(self, request, *args, **kwargs):
+        if not settings.DMOJ_USER_DATA_DOWNLOAD or self.request.profile.mute:
+            raise Http404()
+        return super().dispatch(request, *args, **kwargs)
+
+
+class UserPrepareData(LoginRequiredMixin, UserDataMixin, TitleMixin, FormView):
+    template_name = 'user/prepare-data.html'
+    form_class = DownloadDataForm
+
+    @cached_property
+    def _now(self):
+        return timezone.now()
+
+    @cached_property
+    def can_reprepare_data(self):
+        return (
+            self.request.profile.data_last_downloaded is None or
+            self.request.profile.data_last_downloaded + settings.DMOJ_USER_DATA_RATELIMIT < self._now or
+            not os.path.exists(self.data_path)
+        )
+
+    @cached_property
+    def data_cache_key(self):
+        return 'user_data_download_%s' % self.request.profile.id
+
+    def build_task_url(self, status_id):
+        return task_status_url_by_id(
+            status_id, message=_('Preparing your data...'), redirect=reverse('user_prepare_data'),
+        )
+
+    def get_title(self):
+        return _('Download your data')
+
+    def form_valid(self, form):
+        self.request.profile.data_last_downloaded = self._now
+        self.request.profile.save()
+        status = prepare_user_data.delay(self.request.profile.id, json.dumps(form.cleaned_data))
+        cache.set(self.data_cache_key, status.id)
+        return HttpResponseRedirect(self.build_task_url(status.id))
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context['can_reprepare_data'] = self.can_reprepare_data
+        context['can_download_data'] = os.path.exists(self.data_path)
+        status_id = cache.get(self.data_cache_key)
+        status = task_status_by_id(status_id).status if status_id else None
+        context['in_progress_url'] = (
+            self.build_task_url(status_id)
+            if status and status in ('PENDING', 'PROGRESS', 'STARTED')
+            else None
+        )
+        context['ratelimit'] = settings.DMOJ_USER_DATA_RATELIMIT
+
+        if not self.can_reprepare_data:
+            context['time_until_can_reprepare'] = (
+                settings.DMOJ_USER_DATA_RATELIMIT - (self._now - self.request.profile.data_last_downloaded)
+            )
+        return context
+
+    def post(self, request, *args, **kwargs):
+        if not self.can_reprepare_data:
+            raise PermissionDenied()
+        return super().post(request, *args, **kwargs)
+
+
+class UserDownloadData(LoginRequiredMixin, UserDataMixin, View):
+    def get(self, request, *args, **kwargs):
+        if not os.path.exists(self.data_path):
+            raise Http404()
+
+        response = HttpResponse()
+        if hasattr(settings, 'DMOJ_USER_DATA_INTERNAL') and \
```

Comment:
We have this `X-Accel-Redirect` hack in at least two places now; its complicated enough that it ought to be in a util file or something.